### PR TITLE
Refactor frontend main page and badge-example code

### PR DESCRIPTION
**Pull request by paulmelnikow**
- The goal of this PR is:
    - Consume the new service-definition format
    - Make the frontend more readable.
- Behavior changes:
    - I changed the **Image** field in the markup modal to show only the path.
    - I added another click-to-select field below that shows the complete URL.
    - This made it easier to suppress the live badge preview while it contains placeholders like `:user` or `:gem`, a minor tweak discussed at https://github.com/badges/shields/issues/2427#issuecomment-442972100.
- I did not deliberately address performance, though I ripped out a bunch of anonymous functions and avoided re-filtering all the examples by category on every render, which I expect will not hurt. I haven't really tested this on a mobile connection and it'd be worth doing that.
- It would be great to have some tests of the components, though getting started with that seemed like a big project and I did not want to make this any larger than it already is.

It's a medium-sized refactor:

1. Replace `BadgeExamples`, `Category` and `Badge` component with a completely rewritten `BadgeExamples` component which renders a table of badges, and `CategoryHeading` and `CategoryHeadings` components.
2. Refactor `ExamplesPage` and `SearchResults` components into a new `Main` component.
3. Rewrite the data flow for `MarkupModal`. Rather than rely on unmounting and remounting the component to copy the badge URL into state, employ the `getDerivedStateFromProps` lifecycle method.
4. Remove `prepareExamples` and `all-badge-examples`.
5. Rewrite the `$suggest` schema to harmonize with the service definition format. It's not backward-compatible which means at deploy time there probably will be 10–20 minutes of downtime on that feature, between the first server deploy and the final gh-pages deploy.  🤷‍♂️ (We could leave the old version in place if it seems worth it.)
6. Added two new functions in `make-badge-url` with tests. I removed _most_ of the uses of the old functions, but there are some in parts of the frontend I didn't touch like the static and dynamic badge generators, and again I didn't want to make this any larger than it already is.
7. Fix a couple bugs in the service-definition export.

### CODE REVIEW

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging debfa4667d846c26b47653952aee7ffba739fd53 into 0a6afb447877e9b9d973146be65cb1b23fe7e588 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-1fe6121f2157fefc47875813be21ae99e5ddab64)

**new alerts:**

* 1 for Unused or undefined state property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 2 alerts** when merging b475529238a804ea04845ecad45507c69226cd0e into 0a6afb447877e9b9d973146be65cb1b23fe7e588 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-eb3a2d753465ff7ac12dd2b9b0d72b8c92ae0552)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Unused or undefined state property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
AFAICT the LGTM warning is a false positive.

screen shot 2018-12-02 at 10 39 03 pm

screen shot 2018-12-02 at 10 40 22 pm

**Comment by paulmelnikow** [comment]
This pull request **introduces 2 alerts** when merging c45fc6e7da8bc5d42d74e897180736a7b451d879 into ec07f23942373930a855ad13c8e1b5f6e1b02052 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-2edc63459b3df11193c277182a248e7c458f3834)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Unused or undefined state property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: frontend/components/main.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      isSearchInProgress: false,
+      isQueryTooShort,
+      searchResults,
+    })
+  }
+
+  searchQueryChanged(query) {
+    /*
+    Add a small delay before showing search results
+    so that we wait until the user has stipped typing
```
Comment (chris48s) [inline frontend/components/main.js]:
```suggestion
    so that we wait until the user has stopped typing
```

this was probably originally my typo, but I just spotted it :)

File: frontend/lib/service-definitions/service-definition-set-helper.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  notDeprecated() {
+    return ServiceDefinitionSetHelper.create(
+      this.definitionData.filter(({ isDeprecated }) => !isDeprecated)
+    )
+  }
+
+  asNative() {
+    return this.definitionData
```
Comment (chris48s) [inline frontend/lib/service-definitions/service-definition-set-helper.js]: I find this method name a bit confusing

File: frontend/components/markup-modal.js [hunk 3]
```diff
@@ -2,86 +2,163 @@ import React from 'react'
...
   static propTypes = {
-    example: PropTypes.shape({
-      title: PropTypes.string.isRequired,
-      exampleUrl: PropTypes.string,
-      previewUrl: PropTypes.string,
-      urlPattern: PropTypes.string,
-      documentation: PropTypes.string,
-      link: PropTypes.string,
-    }),
+    example: PropTypes.object,
```
Comment (chris48s) [inline frontend/components/markup-modal.js]: Are we losing some useful validation here?

File: frontend/components/markup-modal.js [hunk 4]
```diff
@@ -160,29 +237,8 @@ export default class MarkupModal extends React.Component {
...
-          </p>
-          <p>
-            AsciiDoc&nbsp;
-            <ClickToSelect>
-              <input className="code clickable" readOnly value={asciiDoc} />
-            </ClickToSelect>
-          </p>
-          {this.renderDocumentation()}
+          {this.isOpen && this.renderMarkup()}
+          {this.isOpen && this.renderDocumentation()}
```
Comment (chris48s) [inline frontend/components/markup-modal.js]: It looks like you've added code to render the documentation but if I look at some examples on https://shields-staging-pr-2441.herokuapp.com/ which include a documentation key, it isn't rendering

File: frontend/lib/service-definitions/service-definition-set-helper.js (hunk 2 above)
Comment (paulmelnikow) [reply frontend/lib/service-definitions/service-definition-set-helper.js]: Would `toArray()` be better?

File: frontend/components/markup-modal.js (hunk 3 above)
Comment (paulmelnikow) [reply frontend/components/markup-modal.js]: Part of what makes prop types useful is that they gives a single point of reference on the props schema. Since the service definition data is passed straight through to the components – this is one of the `example` keys from it – the service definition schema can be used as a reference.

It seems like a lot of work to maintain this schema in two places, in two incompatible formats.

Prop types also so not prevent components from incorrectly referencing props, or referencing props that don't exist. Related to that… in #1272 @Daniel15 suggested adding typing to the repo. This changeset took many hours to produce, and most of the reason is that the types were changing drastically, from the top all the way down. Type checking seems pretty essential for frontends that can be refactored with ease. I suspect adding TypeScript might make the frontend easier for the average potential frontend contributor, despite the added learning curve. 

I continue to think this is not as pressing an issue for the server code and definitely think it will deter some contributors. JavaScript, even with ES6, is a least-common-denominator kind of language. It's pretty easy for anyone to read, and even easier to write.

I'm wanting to make significant improvements to the frontend, and I'm thinking we may need to recruit some frontend contributors in a targeted way in order to do that. It's on my mind to create a front end dev experience that's as good as the back end dev experience. Of course, we also need to spend some time on the design (#1899) so that can happen in an organized way.

But that's all an aside. TypeScript is hardly pressing.

tldr; I think the validation is less important than the documentation. Maybe adding a link to the schema documentation would be sufficient.

File: frontend/components/markup-modal.js (hunk 4 above)
Comment (paulmelnikow) [reply frontend/components/markup-modal.js]: Fixed in 1a8feaf. Ironically enough, had I specced out the input as you suggested in https://github.com/badges/shields/pull/2441#discussion_r240007528, that would have surfaced the problem.

**Comment by paulmelnikow** [comment]
This pull request **introduces 2 alerts** when merging 44095635f667cf6c23b5114085132e900de69a3b into 8a8311d931b9949ae9fdff72c96abb60019825da - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-0c76a9d20b1d3264fab6bfbf55b2c5d177025552)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Unused or undefined state property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: frontend/lib/service-definitions/service-definition-set-helper.js (hunk 2 above)
Comment (chris48s) [reply frontend/lib/service-definitions/service-definition-set-helper.js]: yeah, that seems more descriptive

File: frontend/components/markup-modal.js (hunk 3 above)
Comment (chris48s) [reply frontend/components/markup-modal.js]: There's quite a lot of stuff there, but I think the key point is

> Prop types also so not prevent components from incorrectly referencing props, or referencing props that don't exist

:+1:

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging 74741f70621a53614b5db927869c064ada8d7ff0 into 8a8311d931b9949ae9fdff72c96abb60019825da - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-5c952a564b8057592d3a6196f9138bfeb48e1842)

**new alerts:**

* 1 for Unused or undefined state property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: frontend/components/markup-modal.js (hunk 3 above)
Comment (Daniel15) [reply frontend/components/markup-modal.js]: > Part of what makes prop types useful is that they gives a single point of reference on the props schema. Since the service definition data is passed straight through to the components – this is one of the example keys from it – the service definition schema can be used as a reference.

You can reuse prop types; you don't have to redefine them everywhere. You could make a module that exports the prop types
```
module.exports = {
  Example: PropTypes.shape({
    title: PropTypes.string.isRequired,
    exampleUrl: PropTypes.string,
    previewUrl: PropTypes.string,
    urlPattern: PropTypes.string,
    documentation: PropTypes.string,
    link: PropTypes.string,
  }),
};
```

Then you could do `example: MyShapes.Example` or whatever here.

That's pretty similar to how you'd do this with Flow or TypeScript - Define the type in some reusable place, and reuse it everywhere.

Comment (paulmelnikow) [reply frontend/components/markup-modal.js]: Hi! Right, thanks for that suggestion. I have done that in some places. The issue here is different: [the schema is defined using Joi](https://github.com/badges/shields/blob/master/services/service-definitions.js).

That's the validation framework we're using for the JSON responses we get back, which we adopted for runtime use without a ton of consideration (beyond it being integrated into our API testing framework).

That said, last I looked, I did not find a library that could do runtime validation of plain JS objects using TypeScript types.

There do seem to be some attempts to use Joi with React. I didn't dig too far into [the most promising example](https://www.npmjs.com/package/react-joi-validation), though I also wouldn't want to ship Joi to the frontend.

I wouldn't mind using it in dev, though it seems like it would take some effort to configure webpack to accomplish that.
